udata/core/dataset/forms.py

```diff
@@ -32,46 +32,51 @@ def normalize_format(data):
         return data.strip().lower()
 
 
-def enforce_allowed_schemas(form, field):
-    schema = field.data
-    if schema:
-        allowed_schemas = [s['id'] for s in ResourceSchema.objects()]
+class SchemaForm(ModelForm):
+    model_class = Schema
+    url = fields.URLField(_('URL of the schema'))
+    name = fields.StringField(_('Name of the schema'))
+    version = fields.StringField(_('Version of the schema'))
 
-        if not bool('name' in schema) ^ bool('url' in schema):
-            raise validators.ValidationError(_('Schema must have at least a name or an url. Having both is not allowed.'))
+    def validate_url(form, field):
+        if not field.data: 
+            if form.name.data:
+                return
+            raise validators.ValidationError(_('URL is required when name is missing.'))
 
-        if 'url' in schema:
-            try:
-                validate_url(schema.get('url'))
-            except ValidationError:
-                raise validators.ValidationError(_('Provided URL is not valid.'))
+        if form.name.data:
+            raise validators.ValidationError(_('Having both name and URL is not allowed.'))
 
-        if 'name' in schema and schema.get('name') not in allowed_schemas:
+    def validate_name(form, field):
```

Comment:
As discussed, we may want to move this validation logic at the model level, but call it explicitly from this form validation?

Reply:
Done